gist/gist

```diff
@@ -0,0 +1,812 @@
+_set_gist() {
+  files=()
+  public=True
+  while [[ -n "$*" ]]; do case $1 in
+    -d | --desc)
+      description="$2"
+      shift; shift;;
+    -f | --file)
+      filename="$2"
+      shift; shift;;
+    -p)
+      public=False
+      shift;;
+    *)
+      files+=($1)
+      shift;;
+    esac
+  done
+  ls "${files[@]}" > /dev/null || return 1
+}
+
+# Let user type the content of gist before setting filename
+_new_file() {
+  tmp_file=$(mktemp)
+  if [[ -z $INPUT ]]; then 
+    echo "Type a gist. <Ctrl-C> to cancel, <Ctrl-D> when done" > /dev/tty
+    cat > "$tmp_file"
+  else
+    echo "$INPUT" > "$tmp_file"
+  fi
+
+  echo > /dev/tty
+  [[ -z $1 ]] && read -r -p 'Type file name: ' filename < /dev/tty
+  mv "$tmp_file" /tmp/"$filename"
+  echo /tmp/"$filename"
+}
+
+# Parse JSON object of a single gist 
+_gist_body(){
+  _process_json "
+import os.path
+files_json = {}
+files = sys.stdin.readline().split()
+description = sys.stdin.readline().replace('\n','')
+for file in files:
+    with open(file, 'r') as f:
+        files_json[os.path.basename(file)] = {'content': f.read()}
+print(json.dumps({'public': $public, 'files': files_json, 'description': description}))
+  "
+}
+
+# Create a new gist with files. If success, also update index file and clone the repo
+_create_gist() {
+  _set_gist "$@" || return 1
+  [[ -z ${files[*]} ]] && files+=($(_new_file "$filename"))
+  [[ -z $description ]] && read -r -p 'Type description: ' description < /dev/tty
+
+  echo 'Creating a new gist...'
+  http_data=$(mktemp)
+  echo -e "${files[*]}\n$description" \
+  | _gist_body > "$http_data" \
+  && http_method POST $GITHUB_API/gists \
+  | sed -e '1 s/^/[/; $ s/$/]/' \
+  | _parse_response $(( $(sed -e '/^s/ d' $INDEX | wc -l) +1 )) \
+  | tee -a $INDEX \
+  | cut -d' ' -f2 | sed -E -e 's#.*/##' \
+  | (xargs -I{} git clone "$(_repo_url {})" $folder/{} &> /dev/null &)
+
+  # shellcheck disable=2181
+  if [[ $? -eq 0 ]]; then
+    echo 'Gist is created'
+    INPUT=$(tail -1 $INDEX | cut -d' ' -f1) hint=false _show_list | tail -1
+  else
+    echo 'Failed to create gist'
+  fi
+}
+
+# get the gist description with a given index
+_get_desc() {
+  sed -ne "/^$1 / p" $INDEX | cut -d' ' -f${#INDEX_FORMAT[@]}-
+}
+
+# Update description of a gist
+_edit_gist() {
+  _gist_id "$1" || return 1
+
+  local prefill="$(_get_desc $1)"
+  if [[ -z $2 ]]; then
+    read -e -p 'Edit description: ' -i "$prefill" -r DESC < /dev/tty
+  else
+    DESC="$2"
+  fi
+
+  http_data=$(mktemp)
+  echo '{' \"description\": \""${DESC//\"/\\\"}"\" '}' > "$http_data"
+  new_record=$(http_method PATCH "$GITHUB_API/gists/$GIST_ID" \
+               | sed -e '1 s/^/[/; $ s/$/]/' \
+               | _parse_response "${1#[[:alpha:]]}" )
+  [[ -n $new_record ]] && sed -i'' -E -e "/^$1 / s^.+^$new_record^" $INDEX \
+  && hint=false mark="$1 " _show_list \
+  || echo 'Fail to modify gist description'
+}
+
+# Print helper message
+usage() {
+  sed -E -n -e ' /^$/ q; 7,$ s/^# //p' "$0"
```

Comment:
Can you clarify what is going on here? Most other tools just do 
```bash
  cat <<EOF
Usage goes here text goes here
EOF
```

Reply:
>Can you clarify what is going on here?

Print heading comments inside script. That is, the helper message (But exclude author information)

Compare to other tools, `gist` has too many parameters (sub-commands) and derivative　usages. Also, this script is also a gist on `gist.github.com`. That's why I prefer to put those necessary information at the top of script, so a user can get into it very quickly even it is not downloaded yet, or checking it in an editor. 

```
sed -E -n -e ' /^$/ q; 7,$ s/^# //p' "$0"
```
But specify a fixed line (7) in `sed` may not fit general use case. So I fixed it with pattern. check 4ec969f